Introduce pallet-receipts to keep track of all the receipt-related onchain state

NingLin-P opened the pull request:
close #960 

This PR introduces `pallet-receipt` to keep track of all the receipt-related on-chain states, including:
- `BlockHash`
- `ReceiptHead`
- `OldestReceiptNumber`
- `Receipts`
- `ReceiptVotes`
- `StateRoots`

such that both `pallet-domains` and `pallet-domain-registry` can use this pallet to share a lot of common logic about maintaining receipt-related states.

For the changes this PR introduced, you can see from the commit message, but overall there are no logical changes (except one about maintaining the `BlockHash` for core domains, which I will comment on in more detail below), the code path of maintaining the receipt-related for both system/core domains are same as before.

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](../CONTRIBUTING.md)

NingLin-P commented on domains/client/domain-executor/src/system_bundle_processor.rs:
```diff
@@ -114,7 +114,13 @@ where
+        {
+            // Push the `submit_primary_hash` extrinsic to the end so that it will take effect for the next block
+            extrinsics.push(primary_hash_extrinsic);
+        }
```
This PR maintains the `BlockHash` for core domains by introducing a new `submit_primary_hash` extrinsic and submitting this extrinsic while the system domain process bundles (similar to `submit_core_bundle`). 

The `(primary_number, primary_hash)` used is actually the same as the current implement I think, but this PR only maintains one `BlockHash` state for all the core domains instead of one for each core domain. /cc @liuchengxu

liuchengxu replied:
Using a shared `BlockHash` for all domains is incorrect, let's say a core domain runs into some issue and stalls, but the other domains move forward regularly, once the `BlockHash` is pruned due to the other workings domains reaching the pruning depth, the state of the stalled domain will be corrupted, thus `BlockHash` for each domain is necessary.

NingLin-P commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -247,23 +240,9 @@ mod pallet {
-
-            // TODO: slash the executor accordingly.
+            // FIXME: currently core domain fraud proof will also submit into primary chain and
+            // handled by this extrinsic incorrectly
```
More detail in https://github.com/subspace/subspace/issues/990

liuchengxu replied:
I prefer not to introduce incomplete refactoring in pallet-receipts, it can simply focus on the state maintenance of happy path, and the unhappy path (fraud proof related) can be another separate PR..

liuchengxu reviewed (changes requested):
Some general comments with a quick look:

- Rebase before creating the PR, the manual commit of merging the main is totally unnecessary.
- Generally not force push after creating the PR.
- Run `cargo clippy --all-targets` before creating the PR
- If possible, starting with copy-paste and then adding your modifications can make refactoring easier to review.

Due to the issue I have seen in this PR, I'll have another review once they are addressed.

liuchengxu commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -200,31 +177,15 @@ mod pallet {
+                // missing system domain receipt.
+                .map_err(ExecutionReceiptError::from)
+                .map_err(BundleError::Receipt)
+                .map_err(Error::<T>::from)?;
```
```diff
-                // This error should never happen as the `pre_dispatch_submit_bundle` check ensures no
-                // missing system domain receipt.
-                .map_err(ExecutionReceiptError::from)
-                .map_err(BundleError::Receipt)
-                .map_err(Error::<T>::from)?;
+                .map_err(|err| Error::<T>::Bundle(BundleError::Receipt(err.into())))?;
```

liuchengxu commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -529,20 +407,28 @@ impl<T: Config> Pallet<T> {
+                    if pallet_receipts::Pallet::<T>::primary_hash(
+                        DomainId::SYSTEM,
+                        receipt.primary_number,
+                    ) != receipt.primary_hash
```
Can be more descriptive on the call site ff we can make a small helper function in pallet-receipts for this pattern as it's also repeated several times:

```rust
fn point_to_valid_primary_block(domain_id: DomainId, &receipt: Receipt) -> bool {
    todo!()
}
```
```rust
if !pallet_receipts::Pallet::<T>::point_to_valid_primary_block(DomainId::SYSTEM, &receipt) {
...
```

liuchengxu commented on crates/subspace-runtime/src/lib.rs:
```diff
@@ -476,6 +481,7 @@ construct_runtime!(
         ObjectStore: pallet_object_store = 11,
         Domains: pallet_domains = 12,
         RuntimeConfigs: pallet_runtime_configs = 14,
+        Receipts: pallet_receipts = 15,
```
Better to move `Receipts` before `Domains` in the runtime construction as sometimes the initialization order matters if one pallet is dependent on another.

```
        Receipts: pallet_receipts = 15,
        Domains: pallet_domains = 12,
        RuntimeConfigs: pallet_runtime_configs = 14,
```

liuchengxu commented on crates/pallet-receipts/src/lib.rs:
```diff
@@ -0,0 +1,366 @@
+    }
+
+    /// Track the execution receipts for the domain
+    pub fn track_receipts(
```
Can be easier to review if these pure function abstractions(`track_receipts`, `submit_fraud_proof`, `remove_expired_receipts`) can be included in the first refactoring commit.

liuchengxu commented on crates/pallet-receipts/src/lib.rs:
```diff
@@ -0,0 +1,366 @@
+    }
+
+    /// Submit fraud proof that targetted a given domain
+    pub fn submit_fraud_proof(domain_id: DomainId, fraud_proof: FraudProof) {
```
Something `process_fraud_proof` is somehow nicer to me so that it's distinctive from the Call with same name in pallet-domains.

liuchengxu commented on crates/pallet-domains/Cargo.toml:
```diff
@@ -26,6 +27,7 @@ sp-std = { version = "5.0.0", default-features = false, git = "https://github.co
 sp-io = { version = "7.0.0", git = "https://github.com/subspace/substrate", rev = "6d57dbc639bb3d9460dabeccb063cc6556452535" }
 sp-trie = { version = "7.0.0", git = "https://github.com/subspace/substrate", rev = "6d57dbc639bb3d9460dabeccb063cc6556452535" }
 
+
```
```suggestion
```

Also missing `pallet-receipts/std` below.

liuchengxu reviewed (commented):
Although I haven't thought the fraud-proof part over, this revision is generally more reviewable! 

Not sure when the testnet will be reset, I guess we'd better to defer this PR to be merged until the next network reset so that we don't have to be worried about the potential runtime migration.

NingLin-P committed:
move CoreDomainTracker to sp-domain and extract helper functions track_receipts, process_fraud_proof and remove_expired_receipts from pallet-domains

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
extract pallet-receipts from pallet-domains and use it in pallet-domains

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
use pallet-receipts in domain-register

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Process the core domain fraud proof.
The logic of validating and processing the core domain fraud proof
is the same as the system domain.

Signed-off-by: linning <linningde25@gmail.com>

liuchengxu commented on crates/pallet-receipts/src/lib.rs:
```diff
@@ -0,0 +1,370 @@
+
+//! # Pallet Receipt track all the execution receipt related onchain state for both primary
+//! chain (tracking the system domain execution receipt) and system domain (tracking the core
+//! domains execution receipt)
```
```suggestion
//! # Receipts Pallet
//! This pallet provides the general common settlement functions needed by the consensus chain
//! and system domain, which mainly includes tracking the receipts and handling the fraud proofs
//! from the chain they secure.
```

liuchengxu commented on domains/pallets/domain-registry/src/lib.rs:
```diff
@@ -1109,7 +1032,30 @@ impl<T: Config> Pallet<T> {
 
     // TODO: Verify fraud_proof.
-    fn validate_fraud_proof(_fraud_proof: &FraudProof) -> Result<(), Error<T>> {
+    fn validate_fraud_proof(fraud_proof: &FraudProof) -> Result<(), Error<T>> {
```
This should be identical to the one in pallet-domains, we can extract it into pallet-receipts as well.

liuchengxu reviewed (commented):
A few last comments, and then I think we can finally merge this PR as the new network is on the plan.

NingLin-P committed:
Apply review suggestion

Signed-off-by: linning <linningde25@gmail.com>

liuchengxu reviewed (approved):
LGTM 🎉
